[9.2] Provide a theme through an app

phisch opened the pull request:
<!--
Thanks for submitting a change to ownCloud!

This is the bug tracker for the Server component. Find other components at https://github.com/owncloud/core/blob/master/.github/CONTRIBUTING.md#guidelines

For fixing potential security issues please see https://owncloud.org/security/

To make it possible for us to get your change reviewed and merged please fill out below information carefully.

Please note that any kind of change first has to be submitted to the master branch which holds the next major version of ownCloud.

We will carefully discuss if your change can or has to be backported to stable branches.
-->
## Description

This allows users to install apps that provide a theme. If an app enables itself as app-theme in its `app.php`:

```
/** @var Theme\ThemeService $themeService */
$themeService = \OC::$server->query('ThemeService');
$themeService->setAppTheme('theme-owncloud');
```

the app will be used as theme. If no app claims to be a theme, the existing logic will continue to work. That means, OC will look for the `theme` key in `config/config.php` and will use it. OC will still look for `themes/default` if no theme is configured.
## Related Issue
#20200
# Breaking changes

The documentation (https://doc.owncloud.org/server/9.0/developer_manual/core/theming.html) describes, that modifying an apps translation files would require a file in `theme/apps/app-name/l10n/language.js`. Modifying an apps template on the other hand would require a file in `theme/app-name/templates/template.php`.

This inconsistency has been removed. Templates will be located just like translations under `theme/apps/app-name/template/...`.

Theming an App now requires the corresponding files to be located in an "app" folder.

@DeepDiver1975, @PVince81, @butonic: what do you think about getting rid of this unnecessary complexity?
## Types of changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)
## Checklist:

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->

<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

DeepDiver1975 committed:
Adding Open Graph Support for public shares http://ogp.me/

DeepDiver1975 committed:
In case no preview is available we display the ownCloud logo

DeepDiver1975 committed:
Fix unit test

DeepDiver1975 committed:
Use correct name for background jobs

fixes #26202

Commit:
Add using casing check/fix for initMountPoints

Commit:
Fix post_unshareFromSelf hook parameter format

When unsharing from self in a group share situation, the share items
passed to the post_unshareFromSelf hook were using the wrong format in
which the attribute names (ex: "share_type") have non camel-case format.

This fix makes sure that in group sharing case we use the correct
format. It looks like the code was already producing it but in
array_merge it was not using it and adding the unprocessed one.

DeepDiver1975 committed:
Sanitize length headers when validating quota

Commit:
added dumb Theme value-object

Commit:
added dumb Theme value-object

Commit:
added dumb Theme value-object

Commit:
capsuled theme directory logic and value object creation in ThemeService

Commit:
capsuled theme directory logic and value object creation in ThemeService

Commit:
capsuled theme directory logic and value object creation in ThemeService

Commit:
capsuled theme directory logic and value object creation in ThemeService

Commit:
registered ThemeService as DI service

Commit:
registered ThemeService as DI service

Commit:
registered ThemeService as DI service

Commit:
registered ThemeService as DI service

Commit:
getTheme should return the Theme value-object

Commit:
getTheme should return the Theme value-object

Commit:
getTheme should return the Theme value-object

Commit:
use value-object to search for defaults.php

Commit:
use value-object to search for defaults.php

Commit:
use value-object to search for defaults.php

Commit:
pass theme value-object, use value-object directory for template loading

Commit:
pass theme value-object, use value-object directory for template loading

Commit:
pass theme value-object, use value-object directory for template loading

Commit:
inject theme object into resource locators and use its directory to search for scripts and styles

Commit:
inject theme object into resource locators and use its directory to search for scripts and styles

Commit:
inject theme object into resource locators and use its directory to search for scripts and styles

Commit:
use theme value-object to look for images

Commit:
use theme value-object to look for images

Commit:
use theme value-object to look for images

SergioBertolinSG committed:
Added functions to check etag of elements

ppaysant committed:
console.log now returns 1 instead of 0 on all errors

Commit:
added vfsStream to use a virtual filesystem for tests

Commit:
added vfsStream to use a virtual filesystem for tests

Commit:
added vfsStream to use a virtual filesystem for tests

Commit:
tested value object

Commit:
tested value object

Commit:
tested value object

Commit:
injectable default theme directory to make this testable without reflection, fixed bug where empty theme name and non existing default theme would generate a false theme object, tested ThemeService

Commit:
injectable default theme directory to make this testable without reflection, fixed bug where empty theme name and non existing default theme would generate a false theme object, tested ThemeService

Commit:
injectable default theme directory to make this testable without reflection, fixed bug where empty theme name and non existing default theme would generate a false theme object, tested ThemeService

Commit:
applied correct code styling

Commit:
applied correct code styling

Commit:
applied correct code styling

Commit:
applied correct code styling

DeepDiver1975 commented:
conflicting - please rebase - THX @PhilippSchaffrath

PhilippSchaffrath commented:
The tests fail because vfsStream is missing. Does jenkins install require-dev composer dependencies?

Commit:
[tx-robot] updated from transifex

DeepDiver1975 commented:
I'm the makefile there are targets to execute the tests. I guess there is composer call missing to install development dependencies

butonic committed:
log error when setting timezone to UTC fails (#26354)

Commit:
Merge remote-tracking branch 'origin/feature-app-theme' into feature-app-theme

# Conflicts:
#	lib/private/Server.php

SergioBertolinSG committed:
Changed description of variable

Commit:
feature endorse password for share links

works like "enforce password protection", but let the
user optionally remove the password protection after the
password is set.

butonic commented:
> @DeepDiver1975, @PVince81, @butonic: what do you think about getting rid of this unnecessary complexity?

+1

Commit:
fixed bug where wrong app-name was used

PhilippSchaffrath commented:
Example Theme can be found here: https://github.com/PhilippSchaffrath/oc-app-theme

Just clone it into your apps directory and enable the app.

modified the following things as an example:
- login background colors
- header backround color
- file app sidebar and file list templates
- application title
- german translation for file "download"

SergioBertolinSG committed:
Added test case about issue 26346

PVince81 commented:
> I'm the makefile there are targets to execute the tests. I guess there is composer call missing to install development dependencies

The call is already there. Maybe Jenkins needs a quick "make clean".

If "make test-php" doesn't re-fetch the dev deps, it means that on this test env for some reason the timestamp of "lib/composer" is newer than the timestamp of "composer.json" or "composer.lock".

Maybe bumping the timestamp of composer.json might work ?

PVince81 commented:
Looks like it did run the command:
```
+ make test-php TEST_DATABASE=sqlite
php build/composer.phar install --dev
You are using the deprecated option "dev". Dev packages are installed by default now.
Loading composer repositories with package information
Installing dependencies (including require-dev) from lock file
Warning: The lock file is not up to date with the latest changes in composer.json. You may be getting outdated dependencies. Run update to update them.
Nothing to install or update
```

Commit:
updated composer.lock file

Commit:
updated composer.lock file

Commit:
composer update finally includes vfsStream

Commit:
composer update finally includes vfsStream

Commit:
reverted dependency update, only adding vfsstream as new dev-dependency

Commit:
reverted dependency update, only adding vfsstream as new dev-dependency

SergioBertolinSG committed:
Merge pull request #26370 from owncloud/integration_tests_adding_etag_support

Added functions to check etag of elements

Commit:
Merge pull request #26372 from ppaysant/occ__should_returns_1_when_error

console.log now returns 1 instead of 0 on all errors

Commit:
Hide "Add group" for subadmins

Commit:
fixed tests which are not skipped anymore since vfsstream is available now, removed check for vfs library

Commit:
fixed tests which are not skipped anymore since vfsstream is available now, removed check for vfs library

Commit:
removing accidentally committed restriction :/

Commit:
removing accidentally committed restriction :/

DeepDiver1975 committed:
Move /cloud/user and /cloud/capabilities over to AppFramework

Commit:
Remove checks whether OC is running on Windows pt. 2

Commit:
Get rid of very old oc:// stream wrapper (#26381)

karakayasemi committed:
Update file.php post hook calls

Logicaly, postHooks should emit after touch. For chunking file it is already emitting after touch.

Commit:
[tx-robot] updated from transifex

Commit:
endorseLinkPassword changed to enableLinkPasswordByDefault

Commit:
Merge remote-tracking branch 'origin/master' into endorse-password-for-share-link

Commit:
Merge pull request #26380 from owncloud/subadmin-removeaddgroup

Hide "Add group" for subadmins

Commit:
let the user set an empty password for share link

Commit:
Merge pull request #26385 from RealRancor/windows_part2

Remove checks whether OC is running on Windows pt. 2

Commit:
Fix initMountPoints to set usersSetup earlier

This is needed because in some cases like LDAP, the user manager itself
might trigger avatar updates which would internally also call
initMountPoints with the same user. This could cause the same user to
be setup twice, and in some sharing situations could cause recursive
deduplication of shares by adding "(2)" every time.

Commit:
label changed to better match workflow

Commit:
Merge pull request #26376 from tbenk/endorse-password-for-share-link

feature endorse password for share links

Commit:
Merge pull request #26366 from owncloud/computer-quota-properly-for-non-chunking

Sanitize length headers when validating quota

butonic commented:
IMO we should introduce and use a new app type 'theme'. Then we don't need an app.php for a theme but can use 
```
	<types>
		<theme/>
	</types>
```
in the info.xml

- [ ] find the right place where we should load the `theme` apps
- [ ] add `theme` as a valid filetype

PhilippSchaffrath commented:
I thought about exactly that too! It's on my todo now!

Commit:
Merge pull request #26391 from owncloud/fix-unsharefromself-hookvalues

Fix post_unshareFromSelf hook parameter format

Commit:
[tx-robot] updated from transifex

Commit:
Change getAvatarFolder to accept user object

This is to make sure we don't get into user id casing issues.

butonic committed:
introduce callForSeenUsers and countSeenUsers (#26361)

* introduce callForSeenUsers and countSeenUsers

* add tests

* oracle should support not null on clob

* since 9.2.0

DeepDiver1975 committed:
Set version to 0.2.7 to trigger setup of background job

SergioBertolinSG committed:
Get more potential reviewers (#26405)

Commit:
Merge pull request #26398 from owncloud/initmountpoints-userid-casing

Add using casing check/fix for initMountPoints

Commit:
Merge pull request #26387 from karakayasemi/patch-2

Update file.php post hook calls

Commit:
check for `theme` type and enable app as theme if it exists

Commit:
check for `theme` type and enable app as theme if it exists

PhilippSchaffrath commented:
Last commit checks if the app to load as the `theme` type, if so it registers it as such in the ThemeService. This makes the `app.php` obsolete. This works like @butonic described it.

Although that's just a small change, i am having a lot of trouble testing it because of the static nature of this class. Do you guys have a smart idea on how to test the last commit?

DeepDiver1975 commented:
Testing is not really possible with this legacy class.

Commit:
Merge pull request #26203 from owncloud/DeepDiver1975-patch-1

Use correct name for background jobs

Commit:
Merge pull request #26411 from owncloud/userssetup-check-order

[master] Fix initMountPoints to set usersSetup earlier

DeepDiver1975 committed:
Ensure $commands being an array - fixes #26073

Commit:
[tx-robot] updated from transifex

Commit:
Merge pull request #26423 from owncloud/fix-dependency-analyser-commands-array

Ensure $commands being an array

Commit:
Merge pull request #26382 from owncloud/move-more-ocs

Move /cloud/user and /cloud/capabilities over to AppFramework

davitol committed:
Homogenize personal page errors

Fixes https://github.com/owncloud/core/issues/26428

butonic committed:
add upgrade command before repair, handle NeedsUpgradeExcaption better

Commit:
Merge pull request #26429 from owncloud/PersonalPwdErr

Homogenize personal page errors

Commit:
Merge pull request #26430 from owncloud/needsupgradeexception

add upgrade command before repair, handle NeedsUpgradeExcaption better

PVince81 commented:
What happens if multiple theme apps are enabled at the same time ? The last one alphabetically will win ?

PVince81 commented:
> Theming an App now requires the corresponding files to be located in an "app" folder.
> what do you think about getting rid of this unnecessary complexity?

Which complexity exactly ?

Will this use case still work with your app ? => https://github.com/owncloud/apps/pull/2129

DeepDiver1975 reviewed (commented):
> What happens if multiple theme apps are enabled at the same time ? The last one alphabetically will win ?

exactly

DeepDiver1975 commented on lib/private/Theme/ThemeService.php:
```diff
@@ -0,0 +1,88 @@
+
+namespace OC\Theme;
+
+class ThemeService {
```
not needed anymore - theme is define via info.xml

DeepDiver1975 commented:
> Will this use case still work with your app ? => owncloud/apps#2129

I guess not

DeepDiver1975 commented:
You like super heros? Test this theme - https://github.com/DeepDiver1975/marvel-theme

DeepDiver1975 committed:
Adding capability about new chunking - refs #25682 (#26368)

DeepDiver1975 committed:
Delete compatibility.js

Compatibility for older browsers is no longer required

Commit:
Merge pull request #25579 from owncloud/facebook-preview-public-links

[9.2] Adding Open Graph Support for public shares http://ogp.me/

Commit:
Updated .mention-bot (#26435)

Commit:
Remove loading compatibility.js

Commit:
Cleanup more third party stuff when building dist dir

Commit:
Don't auto-trust servers by default when creating remote shares (#26424)

This makes sure that only admins who really want this feature must
opt-in instead of opt-out.

Commit:
[tx-robot] updated from transifex

butonic commented:
@PhilippSchaffrath can you rebase? I seem to run into https://github.com/owncloud/core/issues/26401 when trying to upgrade to this branch. Has been fixed in https://github.com/owncloud/core/pull/26430

Commit:
[tx-robot] updated from transifex

Commit:
Merge pull request #26445 from owncloud/make-dist-delete-composer-bin

Cleanup more third party stuff when building dist dir

Commit:
Merge pull request #26442 from owncloud/DeepDiver1975-patch-2

Delete compatibility.js

Commit:
[tx-robot] updated from transifex

Commit:
Fix misleading SSL/TLS SMTP email configuration (#26447)

Commit:
Merge remote-tracking branch 'origin/feature-app-theme' into feature-app-theme

PhilippSchaffrath commented:
Since the rebase somehow messed this branch up big time, i figured creating a new branch/pr by cherry-picking would be faster then trying to fix this branch.

New PR: #26462

DeepDiver1975 commented:
@PhilippSchaffrath please delete the branch as soon as you no longer need it - THX

PhilippSchaffrath commented:
@DeepDiver1975 will do that as soon as #26462 checks are green, can you re-run the scruntinizer inspections please? It seems like they failed for no reason.
